### DISCUSSION TITLE
Hierarchical frameworks & integration with OpenAI alternatives

### DISCUSSION THREAD

**Discussion by CognitiveCodes** [langroid/langroid#278] [Ideas]
Hello! I'm working on a similar project as some weird form of a hobby (non-professionally). The main differences are that I'm more focused on creating a multi-agent network with a hierarchical structure - where a websocket server is working as the 'brain' which coordinates work of multiple agents connected to it as clients.

https://github.com/CognitiveCodes/NeuralGPT/blob/main/Chat-center/ServerV2.py
![sche](https://github.com/langroid/langroid/assets/133844350/2e28217e-9b24-4938-9ae7-13ee242a2b2d)
![001](https://github.com/langroid/langroid/assets/133844350/8cb035f7-e638-4e80-b7f9-146a155eacfa)

some clients as examples:
https://github.com/CognitiveCodes/NeuralGPT/blob/main/Chat-center/Agent1.py
https://github.com/CognitiveCodes/NeuralGPT/blob/main/Chat-center/TkDocsBot.py https://github.com/CognitiveCodes/NeuralGPT/blob/main/Chat-center/Docsbotport.html https://github.com/CognitiveCodes/NeuralGPT/blob/main/Chat-center/StarCoderOK.js

Besides that I try to avoid using paid OpenAI services as much as possible and instead try to integrate as many 3rd party sources, apps and tools as I can. Right now I'm using mainly Llama2 API provided by Fireworks platform - https://app.fireworks.ai/
I was wondering if by any chance you might be interested in some form, of cooperation?? And if not then I'd like to ask about any possibility of replacing OpenAI API with Fireworks...?

**Comment by pchalasani** [comment]
Hi, thanks for sharing your project, it looks interesting, and I will have to see more closely to understand it better.

Regarding non-openAI models, I'm not familiar with fireworks, but we currently support 100s of LLM providers via the excellent [LitelLLM](https://github.com/BerriAI/litellm) library which provides an OpenAI-like API to many LLM APIs. See the guide [here](https://langroid.github.io/langroid/tutorials/non-openai-llms/):
